### PR TITLE
Add admin endpoints for administering holds

### PR DESCRIPTION
**Pull request by AxonC**
- Adds routes to GET Holds of Navaid, POST new Holds for Navaid, PUT & DELETE existing holds for a Navaid
- Adds documentation for new endpoints
- Adds new model observer to touch hold dependencies

I will address hold restrictions and deemed separated holds in a separate Pull Request.

### CODE REVIEW

File: app/Http/Controllers/Admin/HoldAdminController.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    private const HOLD_NOT_IN_NAVAID_RESPONSE = ['message' => 'Hold not associated with Navaid.'];
+    /**
+     * Get a list of holds for a specified Navaid.
+     *
+     * @param Navaid $navaid
+     * @return JsonResponse
+     */
+    public function getHolds(Navaid $navaid) : JsonResponse
+    {
+        $holds = $navaid->load(['holds'])->holds;
```
Comment (AndyTWF) [inline app/Http/Controllers/Admin/HoldAdminController.php:24]: Is is possible to use dot notation here to load them all at once, e.g. `$holds = $navaid->load(['holds', 'holds.restrictions'])->holds;`

File: app/Http/Controllers/Admin/HoldAdminController.php [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+    /**
+     * Checks whether the description is already used for
+     * the holds of a specified navaid.
+     *
+     * @param Navaid $navaid
+     * @param string $description
+     * @return boolean
+     */
+    private function doesDescriptionAlreadyExist(Navaid $navaid, string $description): bool
```
Comment (AndyTWF) [inline app/Http/Controllers/Admin/HoldAdminController.php:121]: Would this be better named `descriptionAlreadyExists`?

File: app/Http/Controllers/Admin/HoldAdminController.php [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+     * Checks whether the given Hold has the parent
+     * of the specified Navaid.
+     *
+     * @param Navaid $navaid
+     * @param Hold $hold
+     * @return boolean
+     */
+    private function checkHoldBelongsToNavaid(Navaid $navaid, Hold $hold): bool
+    {
+        return $navaid->id == $hold->navaid_id;
```
Comment (AndyTWF) [inline app/Http/Controllers/Admin/HoldAdminController.php:138]: Worth a `===` here?

File: app/Http/Controllers/Admin/HoldAdminController.php [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+     * Checks whether the description is already used for
+     * the holds of a specified navaid.
+     *
+     * @param Navaid $navaid
+     * @param string $description
+     * @return boolean
+     */
+    private function doesDescriptionAlreadyExist(Navaid $navaid, string $description): bool
+    {
+        $existingDescriptions = $navaid->holds->pluck('description');
```
Comment (AndyTWF) [inline app/Http/Controllers/Admin/HoldAdminController.php:123]: Could make this a one-liner with something like `$navaid->holds->where('description', $description)->exists();`

File: app/Observers/HoldObserver.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        Log::info("Touching hold dependency because of the the deletion of hold {$hold->id} in navaid {$hold->navaid_id}");
+        $this->touchDependency();
+    }
+
+    /**
+     * Wrapper function to touch the stand dependency.
+     *
+     * @return void
+     */
+    private function touchDependency() : void
```
Comment (AndyTWF) [inline app/Observers/HoldObserver.php:52]: Maybe too early to commit to this refactor, as we're going to be creating many more dependency-related observers, worth creating some abstract classes / interfaces, that save us having to do the same thing for each different model?

File: database/factories/Hold/HoldFactory.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+     *
+     * @return array
+     */
+    public function definition()
+    {
+        return [
+            'navaid_id' => Navaid::factory()->create()->id,
+            'inbound_heading' => $this->faker->numberBetween(0, 360),
+            'minimum_altitude' => $this->faker->numberBetween(7000, 24000),
+            'maximum_altitude' => $this->faker->numberBetween(7000, 24000),
```
Comment (AndyTWF) [inline database/factories/Hold/HoldFactory.php:29]: Strictly speaking, needs to be greater than minimum, if that's possible with Faker

**Review by AndyTWF — CHANGES_REQUESTED** [review]
Just a few questions / comments :)

File: app/Observers/HoldObserver.php (hunk 5 above)
Comment (AxonC) [reply app/Observers/HoldObserver.php:52]: Yeah, I'll have a look into this separately: https://github.com/VATSIM-UK/uk-controller-api/issues/521

File: app/Http/Controllers/Admin/HoldAdminController.php (hunk 4 above)
Comment (AxonC) [reply app/Http/Controllers/Admin/HoldAdminController.php]: https://github.com/VATSIM-UK/uk-controller-api/pull/520/commits/4663772e05a7c8b0c7dd9185947a0d60bba0eb9a

(couldn't quite do exists, the next best thing)

File: app/Http/Controllers/Admin/HoldAdminController.php (hunk 3 above)
Comment (AxonC) [reply app/Http/Controllers/Admin/HoldAdminController.php]: https://github.com/VATSIM-UK/uk-controller-api/pull/520/commits/d9c85d58d1e7efc722546fd620e8c5cae23d37a4

File: app/Http/Controllers/Admin/HoldAdminController.php (hunk 2 above)
Comment (AxonC) [reply app/Http/Controllers/Admin/HoldAdminController.php]: https://github.com/VATSIM-UK/uk-controller-api/commit/4663772e05a7c8b0c7dd9185947a0d60bba0eb9a

(had a hard time naming this in the first place :D)

File: app/Http/Controllers/Admin/HoldAdminController.php (hunk 1 above)
Comment (AxonC) [reply app/Http/Controllers/Admin/HoldAdminController.php]: https://github.com/VATSIM-UK/uk-controller-api/pull/520/commits/2c888cb0c3cc358be8ffedd163376aea2cbcb004

(nice call!)

File: database/factories/Hold/HoldFactory.php (hunk 6 above)
Comment (AxonC) [reply database/factories/Hold/HoldFactory.php:29]: I'm not sure it is sadly, I had a quick look into it. I guess if you're testing something specific, you'd override the attributes when generating the model anyway

Comment (AndyTWF) [reply database/factories/Hold/HoldFactory.php:29]: Good point!

**Review by AndyTWF — DISMISSED** [review]

**Pull request merged**